### ISSUE TITLE
Are websockets supported?

### ISSUE THREAD

**Issue by ecthiender** [graphql-python/gql#42]
GraphQL spec says it is transport agnostic, but most implementations support HTTP and the [apollo protocol](https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md) over websockets.

Looking here: https://github.com/graphql-python/gql/tree/master/gql/transport it does not look like websockets are supported. But I may be wrong.

I need websocket support for my project, looking for a Python client that has support for the same.

**Comment by waterdrop01** [comment]
Websocket support would indeed be great, so we can have subscriptions :)

**Comment by remorses** [comment]
I can’t find any library for graphql subscriptions in python, we need to implement this asap

**Comment by loic001** [comment]
FYI, another good GraphQL python client [here](https://github.com/profusion/sgqlc/issues/26). Websocket not yet implemented but they plan to do so.

**Comment by ecthiender** [comment]
I just published a python GraphQL client library supporting websockets. I immediately needed it for my project so currently it only supports websocket transport. Check it out [here](https://pypi.org/project/py-graphql-client/), and contributions and comments welcome.

[pip package](https://pypi.org/project/py-graphql-client/)https://pypi.org/project/py-graphql-client/
[Source](https://github.com/ecthiender/py-graphql-client)

**Comment by leszekhanusz** [comment]
I created the pull request #70 to have a websockets transport to connect to apollo servers
It needs Python 3.6 and asyncio (so tests are failing ...)

Supports fetching schema from transport using the introspection query
Supports queries, mutations AND subscriptions
Supports connection using a client ssl certificate

Only one request per connection for now

**Comment by leszekhanusz** [comment]
Now it is working with multiple requests per connection.
Supports multiple subscriptions in parallel.

**Comment by Cito** [comment]
@ecthiender This is now available in v3.0.0a0 (see README). Please let us know how it is working for you.

**Comment by ecthiender** [comment]
@Cito umm.. I needed this functionality back when I was looking for it. So I implemented my own library: https://github.com/ecthiender/py-graphql-client

Thanks!

**Comment by jameslewis4891** [comment]
> Now it is working with multiple requests per connection.
> Supports multiple subscriptions in parallel.

Do you have an example of this please. No matter what I try with asyncio making use of code in https://github.com/graphql-python/gql/blob/master/docs/code_examples/websockets_async.py for subscriptions only and the app blocks. I cannot figure out a way of running my gql subscriptions in parrallel aside form the main thread (thats running Flask)

**Comment by leszekhanusz** [comment]
@jameslewis4891 There is [an example in the gql docs](https://gql.readthedocs.io/en/v3.0.0a5/advanced/async_advanced_usage.html), basically you have to create multiple asyncio tasks. Please note that it is using asyncio so it is always using the same Thread.

**Comment by jameslewis4891** [comment]
> @jameslewis4891 There is [an example in the gql docs](https://gql.readthedocs.io/en/v3.0.0a5/advanced/async_advanced_usage.html), basically you have to create multiple asyncio tasks. Please note that it is using asyncio so it is always using the same Thread.

Thanks @leszekhanusz that works great however I am struggling to test this code I am trying to mock the `async with client as session:` line and running into difficulties

My Code

`
class GraphQLSubscriber:

    def get_subscribing_to(self) -> str:
        pass

    def getName(self) -> str:
        pass

    def getQuery(self) -> str:
        pass

    def getHandler(self) -> Callable:
        pass

    def getVariables(self) -> dict:
        pass


def connection_backoff(data: dict):
    log_message('Connection error to GraphQL - Reattempting connection', error=True)


class GraphQLSubscriptionHandler(threading.Thread):

    def __init__(self, config, subscribers):
        super(GraphQLSubscriptionHandler, self).__init__(daemon=True)
        self.subscribers = subscribers
        self.venue_api_url = get_config_value(VENUE_API_URL, config)

    async def execute_subscription(self, session, document, variable_values, handler):
        async for result in session.subscribe(document=document, variable_values=variable_values):
            handler(result)

    @backoff.on_exception(backoff.expo, Exception, max_time=1000, on_backoff=connection_backoff)
    async def init_subscriptions(self):
        transport = WebsocketsTransport(url=self.venue_api_url)
    
        log_message(f'Connecting to graphql server [{self.venue_api_url}')
    
        client = Client(transport=transport, fetch_schema_from_transport=True)
    
        async with client as session:
            tasks = []
    
            for sub in self.subscribers:
                log_message(f'GQL Subscriber [{sub.getName()}] Subscribing to [{sub.get_subscribing_to()}]')
                tasks.append(asyncio.create_task(self.execute_subscription(session, document=gql(sub.getQuery()),
                                                                           variable_values=sub.getVariables(),
                                                                           handler=sub.getHandler())))
            await asyncio.gather(*tasks)

    def run(self):
        asyncio.run(self.init_subscriptions())
`

My test

`
@pytest.mark.asyncio
async def test_should_subscribe(event_loop):
    mock_subscriber = Mock()

    with patch('app.graphql.graphql_subscription_handler.WebsocketsTransport') as gql_websockets_transport_mock:
        with patch('app.graphql.graphql_subscription_handler.Client') as gql_client_mock:

            # arrange
            mock_session = Mock()

            gql_client_instance = gql_client_mock.return_value

            gql_client_instance.blah.return_value = mock_session

            gql_client_instance.__aenter__.return_value = mock_session
            gql_client_instance.side_effect = mock_session

            graphql_subscription_handler = GraphQLSubscriptionHandler(CONFIG, [mock_subscriber])

            # act
            await graphql_subscription_handler.init_subscriptions()

            # assert
            mock_subscriber.getQuery.assert_called_once()
`

I am interested in ensuring that the session has been called with the correct variables from the passed in subscriber

**Comment by leszekhanusz** [comment]
@jameslewis4891 Please open another issue with your problem. This issue here is closed.

**Comment by jameslewis4891** [comment]
@leszekhanusz thanks please find https://github.com/graphql-python/gql/issues/42
